### DISCUSSION TITLE
how to specify apktool build DEX_FILE_MAGIC dex format

### DISCUSSION THREAD

**Discussion by gfanini** [iBotPeaches/Apktool#3465] [Q&A] [answered]
how to specify apktool build DEX_FILE_MAGIC dex format ?  
I am getting this error running the apk on android 8, from a project I need to downgrade from android 9   
E AndroidRuntime: Suppressed: java.io.IOException: Failed to open dex files from /data/app/com.name==/base.apk because: Unrecognized version number in /data/app/com.name==/base.apk: 0 3 9
could it be \apktool-lib\src\main\java\brut\androlib\src\SmaliBuilder.java mApiLevel ?

**Comment by iBotPeaches** [comment, marked as answer]
You want `-api, --api-level <API>` during build.

**Reply by gfanini** [reply]
so if I wanted intentionally to generate that 039 say, what should I write -api 28 ? for android 9
should I need android 8 I think it's api 26 I should use -api 26 ?

**Reply by iBotPeaches** [reply]
I don't believe thats how it works. The API level is instructing what compatibility the smali is to build again. You probably want to use it during disassembly to generate an older version of smali.

Might be wrong on that though - haven't messed with those flags in years. Give it a test and let me know.

**Reply by gfanini** [reply]
the build\apk\classes.dex appears to be created during build, not disassembly, and I see empirically that setting b -api 26 has produced a classes.dex with 038 for android 8
https://source.android.com/docs/core/runtime/dex-format
Support for version 038 of the format was added in the Android 8.0 release.
